File: baekjoon/10451_순열사이클/10451_순열사이클.py
#34
import sys
sys.setrecursionlimit(10**6)
def dfs(graph,visited,start):
    n = graph[start]
    if not visited[n]:
        visited[n] = True
        dfs(graph,visited,n)
        
    return

T = int(input())

for i in range(T):
    N = int(input())
    graph = [0] + list(map(int,input().split()))
    visited = [False] * (N+1)
    cnt = 0

    for next in range(1,N+1):
        if not visited[next]:
            dfs(graph,visited,next)
            cnt += 1

    print(cnt)

File: baekjoon/gold/10451_순열사이클/10451_순열사이클.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

void dfs(int cur,const vector<vector<int>> &graph, vector<bool> &visited,int &cnt)
{
    visited[cur] = true;
    

    for (int next : graph[cur])
    {
        if(!visited[next])
        {
            dfs(next,graph,visited,cnt);
        }
        
        
    }

}

int main()
{
    int T;
    cin >> T;
    vector<int> result(T);
    
    
    for (int i = 0; i < T; i++)
    {
        int N;
        int cnt = 0;
        cin >> N;
        vector<vector<int>> graph(N+1);
        vector<bool> visited(N+1);
        vector<int> num(N+1);
        int tmp = 0;
        for (int j = 1; j <= N; j++)
        {
            cin >> tmp;
            graph[j].push_back(tmp);
        }
            

        // sort(num.begin(),num.end(),num);
        // for (int x = 1; x <= N; x++)
        //     graph[x] = num[x]
        for (int z = 1; z <= N; z++)
        {
            if(!visited[z])
            {
                dfs(z,graph,visited,cnt);
                cnt++;
            }
            
        }

        
        result[i] = cnt;
    }
    for (int i : result)
        cout << i << endl;
}